File: leetcode/0199-binary-tree-right-side-view/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
vector<int> rightSideView(TreeNode* root) {
    if (root == nullptr) {
        return {};
    }

    queue<TreeNode*> queue;
    queue.push(root);

    vector<int> ans;

    while (!queue.empty()) {
        int nodes_in_current_level = queue.size();

        ans.push_back(queue.back()->val);

        for (int i = 0; i < nodes_in_current_level; i++) {
            TreeNode* node = queue.front();
            queue.pop();

            if (node->left) {
                queue.push(node->left);
            }

            if (node->right) {
                queue.push(node->right);
            }
        }
    }

    return ans;
}

File: leetcode/0199-binary-tree-right-side-view/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
  if (!root) {
    return [];
  }

  let queue: TreeNode[] = [root];
  let ans: number[] = [];

  while (queue.length) {
    const nodesInCurrentLevel = queue.length;
    let nextQueue: Array<TreeNode> = [];

    ans.push(queue[nodesInCurrentLevel - 1].val);

    for (let i = 0; i < nodesInCurrentLevel; i++) {
      let node = queue[i];

      if (node.left) {
        nextQueue.push(node.left);
      }

      if (node.right) {
        nextQueue.push(node.right);
      }
    }

    queue = nextQueue;
  }

  return ans;
}
